bundles/org.openhab.binding.playstation/src/main/java/org/openhab/binding/playstation/internal/PS4Handler.java

```diff
@@ -0,0 +1,821 @@
+                        respBuff);
+            }
+        }
+
+    }
+
+    private SocketChannel getConnection() throws IOException {
+        return getConnection(true);
+    }
+
+    private SocketChannel getConnection(boolean requiresLogin) throws IOException {
+        SocketChannel channel = null;
+        SocketChannelHandler handler = socketChannelHandler;
+        if (handler == null || !handler.isChannelOpen()) {
+            try {
+                handler = new SocketChannelHandler();
+                socketChannelHandler = handler;
+            } catch (IOException e) {
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, e.getMessage());
+                throw e;
+            }
+        }
+        channel = handler.getChannel();
+        if (!handler.loggedIn && requiresLogin) {
+            login(channel);
+        }
+        return channel;
+    }
+
+    private void sendPacketToPS4(ByteBuffer packet, SocketChannel channel, boolean encrypted, boolean restartTimeout) {
+        PS4Command cmd = PS4Command.valueOfTag(packet.getInt(4));
+        logger.debug("Sending {} packet.", cmd);
+        try {
+            if (encrypted) {
+                ByteBuffer outPacket = ps4Crypto.encryptPacket(packet);
+                channel.write(outPacket);
+            } else {
+                channel.write(packet);
+            }
+            if (restartTimeout) {
+                setupConnectionTimeout(config.connectionTimeout);
+            }
+        } catch (IOException e) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, e.getMessage());
+            logger.info("Send packet {} exception: {}", cmd, e.getMessage());
+        }
+    }
+
+    private void sendPacketEncrypted(ByteBuffer packet, SocketChannel channel) {
+        sendPacketToPS4(packet, channel, true, true);
+    }
+
+    private void sendPacketEncrypted(ByteBuffer packet) {
+        sendPacketEncrypted(packet, true);
+    }
+
+    private void sendPacketEncrypted(ByteBuffer packet, boolean requiresLogin) {
+        try {
+            SocketChannel channel = getConnection(requiresLogin);
+            sendPacketToPS4(packet, channel, true, requiresLogin);
+        } catch (IOException e) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, e.getMessage());
+            logger.info("Send packet exception: {}", e.getMessage());
+        }
+    }
+
+    private boolean login(SocketChannel channel) {
+        // Send login request
+        ByteBuffer outPacket = PS4PacketHandler.makeLoginPacket(config.userCredential, config.passCode,
+                config.pairingCode);
+        sendPacketEncrypted(outPacket, channel);
+
+        try {
+            SocketChannelHandler handler = socketChannelHandler;
+            if (handler != null && isLinked(CHANNEL_2ND_SCREEN)) {
+                for (int i = 0; i < 20; i++) {
```

Comment:
This may block for a long time. Wouldn't it make more sense to retry after a defined interval?

Reply:
I moved the sending to when login ok is received.